notes/lecture21.tex

```diff
@@ -0,0 +1,192 @@
+\section{Lecture 21: Non-convex constraints part 1}
+Recall that convex minimization refers to minimizing convex functions over convex constraints. Today we will begin to explore minimizing convex functions with non-convex constraints. It is difficult to analyze the impact of ``non-convexity" in general, since that can refer to anything that is not convex, which is a very broad class of problems. So instead, we will focus on solving least squares with sparsity constraints:
+\begin{align*}
+    \min_{\|x\|_0 \leq s} \|Ax-y\|^2_2
+\end{align*}
+for $y \in \mathbb{R}^{n}$, $A \in \mathbb{R}^{n \times d}$, and $x \in \mathbb{R}^{d}$ where $d < n$. We will show that in general even this problem is hard to solve but that for a restricted class of problems there is an efficient convex relaxation.
+
+Least squares with sparsity constraints can be applied to solving compressed sensing and sparse linear regression, which are important in a variety of domains. In compressed sensing, $A$ is a measurement model and $y$ are the measurements of some sparse signal $x$. Compressed sensing is applied to reduce the number of measurements needed for, say, an MRI because by including a sparsity constraint on $x$ we are able to recover the signal $x$ in fewer measurements.
+
+In sparse linear regression, $A$ is the data matrix and $y$ is some outcome variable. The goal of sparse linear regression is to recover a weights $x$ on a sparse set of features that are responsible for the outcome variable. In genetics, $A$ could be the genes of a patient, and $y$ is whether they have a particular disease. Then the goal is to recover a weights $x$ on a sparse set of genes that are predictive of having the disease or not.
+
+When there is no noise in the linear equations, we can simplify the problem to
+\begin{align*}
+    & \min \|x\|_0 \\
+    &  Ax = y
+\end{align*}
+
+\subsection{Hardness}
+Even this simplification is NP-hard, as we will show with a reduction to exact 3-cover, which is NP-complete.
+\begin{definition}
+The \textit{exact cover by 3-sets} problem is given a collection $\{T_i\}$ of 3-element subsets of $[n]$, does there exist an exact cover of $[n]$, a set $z \subseteq [d]$ such that $\cup_{j \in z} T_j = [n]$ and $T_i \cap T_j = \emptyset$ for $j \neq j' \in z$?
+\end{definition}
+
+\begin{definition}
+The support of a vector $x$ is defined as
+\begin{align*}
+\text{supp}(x) = \{i \mid x_i \neq 0 \}.
+\end{align*}
+\end{definition}
+
+\begin{theorem}
+$l_0$-minimization for general $A$ and $y$ is NP-hard.
+\end{theorem}
+\begin{proof}
+Define matrix $A$ as
+\begin{align*}
+    A_{ij} = \begin{cases}
+    1 & \text{if } j \in T_i \\
```

Comment:
I think $j \in T_i$ should be $i \in T_j$?

Reply:
Makes sense. Each set corresponds to a column.